### PR TITLE
#15 - Implementing Axis Selection

### LINKED ISSUE #15: Implement Axis data choosing
Implement an interaction flow for changing the data that is displayed on each axis

### PR DESCRIPTION
**Pull request by qpb948**
Closes #15 

# What was the issue
Implementing selection of data that will represent static XYZ axes.

# What was done and why
Implemented ways to filter out columns that are not integers and a way to get XYZ axes through a column name.
I plan on having the user click a button and have a menu pop up that lists the possible column names that can be used for XYZ axes. Then three text boxes that will be linked to XYZ axes. ~~Also created a max function that will get the max value from the XYZ axes selected so the range is correctly displayed on the 3D axis.~~ 

Added an `AxesSelectionContext` to update and return the state values of XYZ axes. The `SelectAxesMenu` has functions for 3 column selections and to plot those data points on the graph. Also fixed data point rendering to be in scale with the hard-coded axes start points. Interaction flow for changing the data successfully flows from axis selection menu  to `DataAbstractor` layer and `CreateGraphingDataPoints` to displayed on each axis.

NOTE: selecting 2 or more identical columns from the XYZ axes results in no data point rendering on graph. Front-end displays nothing, just need to fix this from the back-end side.

# How it was tested
~~Create units tests ensuring the functionality of the methods created. Plan on creating more unit tests when front end is being implemented.~~

Manually added a batch item to `App.tsx` with hard coded x, y, z column entires and data point coordinates to test axis selection for raw and PCA columns. Performed statistic and PCA calculations to confirm that indexDB stores the corresponding results.

Screen Shot 2024-03-29 at 11 04 18 PM

Screen Shot 2024-03-29 at 11 04 44 PM

Tested front-end with distinct Raw and PCA axes selections. User is successful able to select any 3 columns of Raw or PCA types to plot onto graph in 3D space.

Screen Shot 2024-03-29 at 11 05 21 PM

Screen Shot 2024-03-29 at 11 06 02 PM

### CODE REVIEW

File: my-webxr-app/src/App.tsx [hunk 1]
```diff
@@ -11,6 +11,7 @@ import DataPoint from './components/DataPoint';
 import { PointSelectionProvider } from './contexts/PointSelectionContext';
 import DataPointMenu from './components/DataPointMenu';
 import createPosition from './components/Positions';
+// import SelectAxesColumns from './components/SelectAxesMenu';
```
Comment (matthew-buglass) [inline my-webxr-app/src/App.tsx:14]: Why is this commented out?

File: my-webxr-app/src/components/SelectAxesButton.tsx [hunk 2]
```diff
@@ -0,0 +1,55 @@
+/*
+this was going to be used to try and
+implement the axis selection into the 3d scene
+not completed in any way, was a work in progress
+ */
+
+import { useState } from 'react';
+import { Interactive } from '@react-three/xr';
+import { Text, Html } from '@react-three/drei';
+// import SelectAxesColumns from './SelectAxesMenu';
```
Comment (matthew-buglass) [inline my-webxr-app/src/components/SelectAxesButton.tsx:10]: Why is this commented out?

File: my-webxr-app/src/components/SelectAxesMenu.tsx [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+/*
+  this function creates the dropdowns that will be used to choose the axes for XYZ
+  @params:
+      - label: the label of the dropdown (the text beside it)
+      - id: the id of the dropdown (to classify x, y, or z)
+      - options: this will be the choices that can fill the dropdowns
+      - chosenValue: will be the value chosen for from each dropdown
+  @return: the dropdown menu
+ */
```
Comment (matthew-buglass) [inline my-webxr-app/src/components/SelectAxesMenu.tsx:24]: Pre and post-conditions

File: my-webxr-app/src/components/SelectAxesMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  );
+}
+
+/*
+  this function will allow the user to choose their axis values
+  @params:
+      - dbName: the name of the database to get the data from
+  @return: the dropdown menus
+  (think also needs to return the columns chosen and assign the values to x,y,z of data points)
+ */
```
Comment (matthew-buglass) [inline my-webxr-app/src/components/SelectAxesMenu.tsx:56]: Pre and post-conditions

File: my-webxr-app/src/repository/DbRepository.tsx [hunk 5]
```diff
@@ -128,6 +129,26 @@ export default class DbRepository extends Dexie implements Repository {
...
+         */
+  async getPossibleAxes(): Promise<string[]> {
+    const allColumns = await this.columns.toArray();
+    return getColumnNames(allColumns);
+  }
+
+  /*
+    selectRepresentingColumn gets the columns which will represent the XYZ axes
+    @return an array of the XYZ axes columns
+ */
```
Comment (matthew-buglass) [inline my-webxr-app/src/repository/DbRepository.tsx:144]: Pre and post-conditions

File: my-webxr-app/src/repository/DbRepository.tsx [hunk 6]
```diff
@@ -128,6 +129,26 @@ export default class DbRepository extends Dexie implements Repository {
     return dataPoints;
   }
 
+  /*
+        getPossibleAxes retrieves the column names that contain only integer values
+        @return Promise<string[]>: an array of the column names with integers only
+         */
```
Comment (matthew-buglass) [inline my-webxr-app/src/repository/DbRepository.tsx:135]: Pre and post-conditions

File: my-webxr-app/src/repository/DbRepository.tsx
Comment (matthew-buglass) [inline my-webxr-app/src/repository/DbRepository.tsx:1]: To mock out the behaviour while we are still waiting for the DAL to finish, we can just have there return static value so we can at least see it doing something.

File: my-webxr-app/src/repository/FilterAxis.tsx [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+  return column.values.every((value) => typeof value === 'number' && Number.isInteger(value));
+}
+
+export function getIntColumns(columns: Array<Column>) {
+  return columns.filter(isIntColumn);
+}
+
+export default function getColumnNames(columns: Array<Column>) {
+  return getIntColumns(columns).map((obj) => obj.name);
+}
```
Comment (matthew-buglass) [inline my-webxr-app/src/repository/FilterAxis.tsx:14]: Doc strings. Pre and post-conditions.

File: my-webxr-app/src/repository/GetMaxXYZRange.tsx [hunk 8]
```diff
@@ -0,0 +1,8 @@
+import Column from './Column';
+
+export default function getMaxXYZ(xAxis: Column, yAxis: Column, zAxis: Column) {
+  const xMax = Math.max(...(xAxis.values as number[]).map(Math.abs));
+  const yMax = Math.max(...(yAxis.values as number[]).map(Math.abs));
+  const zMax = Math.max(...(zAxis.values as number[]).map(Math.abs));
+  return Math.max(xMax, yMax, zMax);
+}
```
Comment (matthew-buglass) [inline my-webxr-app/src/repository/GetMaxXYZRange.tsx:8]: Doc string. Pre and post-conditions

File: my-webxr-app/src/utils/GetColumnNames.tsx [hunk 9]
```diff
@@ -0,0 +1,41 @@
+import DbRepository from '../repository/DbRepository';
+import assert from './Assert';
+
+export async function getColumnTitles(dbName: string): Promise<string[]> {
```
Comment (matthew-buglass) [inline my-webxr-app/src/utils/GetColumnNames.tsx:4]: Doc string. Pre and post-conditions

File: my-webxr-app/src/utils/GetColumnNames.tsx [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+    if (!Array.isArray(axesArray)) {
+      throw new Error('DbRepository.getPossibleAxes() did not return an array');
+    }
+    return axesArray;
+  } catch (error) {
+    throw new Error('Error getting array of possible axes');
+  }
+}
+
+export async function setRepresentingColumns(
```
Comment (matthew-buglass) [inline my-webxr-app/src/utils/GetColumnNames.tsx:19]: Doc string. Pre and post-conditions

**Review by matthew-buglass — CHANGES_REQUESTED** [review]
Overall, pretty good. A few notes:
- To full test this and as a interim fix until the adding of data is hooked up, we can have the DAL methods return static values so we can make sure that this part works when we're using it.
- Really need to beef up the pre- and post-conditions and add more docstrings.
- I need to see some descriptions and ideally screenshots of what this looks like while the app is being run.

File: my-webxr-app/src/components/SelectAxesMenu.tsx [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+  this function creates the dropdowns that will be used to choose the axes for XYZ
+  @params:
+      - label: the label of the dropdown (the text beside it)
+      - id: the id of the dropdown (to classify x, y, or z)
+      - options: this will be the choices that can fill the dropdowns
+      - chosenValue: will be the value chosen for from each dropdown
+  @pre-condition: the options should be loaded up with column names
+  @post-condition: chosenValue is updated with the selection
+  @return: the dropdown menu
+ */
```
Comment (matthew-buglass) [inline my-webxr-app/src/components/SelectAxesMenu.tsx:27]: Params should be listed each on their own line

File: my-webxr-app/src/components/SelectAxesMenu.tsx [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+/*
+  this function will allow the user to choose their axis values
+  @params:
+      - dbName: the name of the database to get the data from
+  @pre-condition: dB name cannot be empty, relies on database
+  @post-condition: setting the proper x,y,z coordinates for the data point
+  @return: the dropdown menus
+  (think also needs to return the columns chosen and assign the values to x,y,z of data points)
+ */
```
Comment (matthew-buglass) [inline my-webxr-app/src/components/SelectAxesMenu.tsx:63]: Params should be listed each on their own line

File: my-webxr-app/src/contexts/AxesSelectionContext.tsx [hunk 13]
```diff
@@ -0,0 +1,126 @@
+import React, {
+  createContext, useContext, useMemo, useState,
+} from 'react';
+
+// import * as log4js from "log4js";
```
Comment (matthew-buglass) [inline my-webxr-app/src/contexts/AxesSelectionContext.tsx:5]: Delete

File: my-webxr-app/src/contexts/AxesSelectionContext.tsx [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+ * There is a default value of null when it used outside a AxisSelectionProvider.
+ * Otherwise, it is the selectedAxis state.
+ */
+export const AxesSelectionContext = createContext<AxesSelectionContextType | null>(null);
+
+/**
+ * Create the Context Provider element for the React tree.
+ *
+ * @param children: pass-through for the child elements.
+ */
```
Comment (matthew-buglass) [inline my-webxr-app/src/contexts/AxesSelectionContext.tsx:48]: pre- and post- conditions

File: my-webxr-app/src/contexts/AxesSelectionContext.tsx [hunk 15]
```diff
@@ -0,0 +1,126 @@
...
+
+  const setSelectedXAxis = (
+    newValue: React.SetStateAction<string | null>,
+  ) => {
+    // log4js
+    //   .getLogger()
+    //   .debug(
+    //     "AxesSelectionContext: updating selectedXAxis state to " +
+    //       newValue,
+    //   );
```
Comment (matthew-buglass) [inline my-webxr-app/src/contexts/AxesSelectionContext.tsx:65]: delete

File: my-webxr-app/src/contexts/AxesSelectionContext.tsx [hunk 16]
```diff
@@ -0,0 +1,126 @@
...
+
+  const setSelectedYAxis = (
+    newValue: React.SetStateAction<string | null>,
+  ) => {
+    // log4js
+    //   .getLogger()
+    //   .debug(
+    //     "AxesSelectionContext: updating selectedYAxis state to " +
+    //       newValue,
+    //   );
```
Comment (matthew-buglass) [inline my-webxr-app/src/contexts/AxesSelectionContext.tsx:77]: delete

File: my-webxr-app/src/contexts/AxesSelectionContext.tsx [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+
+  const setSelectedZAxis = (
+    newValue: React.SetStateAction<string | null>,
+  ) => {
+    // log4js
+    //   .getLogger()
+    //   .debug(
+    //     "AxesSelectionContext: updating selectedZAxis state to " +
+    //       newValue,
+    //   );
```
Comment (matthew-buglass) [inline my-webxr-app/src/contexts/AxesSelectionContext.tsx:89]: delete

File: my-webxr-app/src/contexts/AxesSelectionContext.tsx [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+    <AxesSelectionContext.Provider value={value}>
+      {children}
+    </AxesSelectionContext.Provider>
+  );
+}
+
+/**
+ * Provide a type-guaranteed context (not null) for use within components.
+ * Call this function instead of useContext(AxesSelectionContext).
+ */
```
Comment (matthew-buglass) [inline my-webxr-app/src/contexts/AxesSelectionContext.tsx:116]: pre- and post-conditions

**Review by matthew-buglass — CHANGES_REQUESTED** [review]

File: my-webxr-app/src/contexts/AxesSelectionContext.tsx [hunk 19]
```diff
@@ -0,0 +1,112 @@
...
+  );
+}
+
+/**
+ * Provide a type-guaranteed context (not null) for use within components.
+ * Call this function instead of useContext(AxesSelectionContext).
+ * @pre-condition: component needs to be wrapped inside AxesSelectionProvider component
+ * @post-condition: returns AxesSelectionContext value if it exists,
+ *                  else error is thrown if not used within AxesSelectionProvider
+ */
```
Comment (matthew-buglass) [inline my-webxr-app/src/contexts/AxesSelectionContext.tsx:102]: Missing a `@return`

**Review by matthew-buglass — CHANGES_REQUESTED** [review]
Just a few minor syntax things, but otherwise looks great!

File: my-webxr-app/src/components/SelectAxesMenu.tsx [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+  @param _setSelectedZAxis: setter function for selected z-axis
+  @pre-condition: provided setter functions must be
+                  React.Dispatch<React.SetStateAction<string | null>>
+  @post-condition x,y,z values updated with selected values
+  @return: a promise
+*/
+  const handleCompleteSelection = async (
+    _setSelectedXAxis: React.Dispatch<React.SetStateAction<string | null>>,
+    _setSelectedYAxis: React.Dispatch<React.SetStateAction<string | null>>,
+    _setSelectedZAxis: React.Dispatch<React.SetStateAction<string | null>>,
```
[the remark this answers is not in the document]
Comment (Aesha7) [reply my-webxr-app/src/components/SelectAxesMenu.tsx:95]: I needed to create new variable names and setSelectedXAxis and setXAxis were already used.

**Review by matthew-buglass — APPROVED** [review]
Looks great!

**Pull request merged**